### PR TITLE
feat: [WD-14770] Device inheritance simplification

### PR DESCRIPTION
**Pull request by Kxiru**
## Done

- Created new functions processIngeritedDevices and getLxdDefault

Fixes [list issues/bugs if needed]

## QA

1. Run the LXD-UI:
    - On the demo server via the link posted by @webteam-app below. This is only available for PRs created by collaborators of the repo. Ask @mas-who or @edlerd for access.
    - With a local copy of this branch, [build and run as described in the docs](../CONTRIBUTING.md#setting-up-for-development).
2. Perform the following QA steps:
    - [List the steps to QA the new feature(s) or prove that a bug has been resolved]

## Screenshots

[if relevant, include a screenshot or screen capture]

### CODE REVIEW

File: src/util/configInheritance.tsx [hunk 1]
```diff
@@ -249,6 +244,31 @@ export const getInheritedRootStorage = (
...
+      .filter(([key, device]) => filterFn(device) && key !== "root")
+      .map(([key, dev]) => {
+        inheritedDevices.push({
+          key: key,
+          [device]: dev as T,
+          source: `${profile.name} profile`,
+        });
+      });
+  }
+};
```
Comment (edlerd) [inline src/util/configInheritance.tsx:270]: I think the problem is that in this generic function, we can't ensure the returned values are of a specific device type. I see you tried to solve that with a generic type and passing in a filter function. I think that could work, if the filter function is also using the same type `T`as the return value.

But maybe a simpler approach is to have the generic function collect all devices. In the specific functions, we can focus on filtering and mapping to the correct device type. With this, we don't need to pass the filter around, but apply it where it gets defined.

```suggestion

interface InheritedDevice {
  key: string;
  device: LxdDeviceValue;
  source: string;
}

const getInheritedDevices = (
  values: InstanceAndProfileFormValues,
  profiles: LxdProfile[],
): InheritedDevice[] => {
  const result: InheritedDevice[] = [];
  if (values.entityType === "instance") {
    const appliedProfiles = getAppliedProfiles(values, profiles);
    for (const profile of appliedProfiles) {
      Object.entries(profile.devices).map(([key, device]) => {
        result.push({ key, device, source: `${profile.name} profile` });
      });
    }
  }
  return result;
};

export interface InheritedVolume {
  key: string;
  disk: LxdDiskDevice;
  source: string;
}

export const getInheritedVolumes = (
  values: InstanceAndProfileFormValues,
  profiles: LxdProfile[],
): InheritedVolume[] => {
  return getInheritedDevices(values, profiles)
    .filter((item) => isDiskDevice(item.device) && item.key !== "root")
    .map((item) => ({
      ...item,
      disk: item.device as LxdDiskDevice,
    }));
};
```

Comment (Kxiru) [reply src/util/configInheritance.tsx:270]: This is a great solution. I love how you used the spread operator to transform the item object array and also add another property. That's a very cool technique that I will try to keep in mind. Thanks for the implementation :).

File: src/util/configInheritance.tsx [hunk 2]
```diff
@@ -261,23 +258,12 @@ export const getInheritedVolumes = (
...
-            disk: disk as LxdDiskDevice,
-            source: `${profile.name} profile`,
-          });
-        });
-    }
-  }
-
-  return inheritedVolumes;
+  return getInheritedDevices(values, profiles)
+    .filter((item) => isDiskDevice(item.device))
```
[the remark this answers is not in the document]
Comment (Kxiru) [reply src/util/configInheritance.tsx]: Fair enough, thanks David! :)

**Review by edlerd — DISMISSED** [review]
LGTM, just one linter error left, I think the type cast in line 280 is not needed, because isDiskDevice ensures the type already. Though below in line 284 we probably still need it.

**Review by edlerd — APPROVED** [review]

**Pull request merged**